Huggingface Model Deployer

dudeperf3ct opened the pull request:
## Describe changes

I implemented ModelDeployer component to work with Huggingface repos.

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).
- [ ] If my changes require changes to the dashboard, these changes are communicated/requested.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
	- Introduced Huggingface integration for deploying machine learning models using Huggingface's infrastructure.
	- Added support for configuring and managing Huggingface inference endpoints.
	- New deployment functionality for Huggingface models, including creation, update, and stopping of model deployment services.
	- Implemented a pipeline step for continuous deployment with Huggingface Inference Endpoint.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

htahir1 commented:
@coderabbit review

htahir1 commented:
@coderabbitai review

strickvl reviewed (changes requested):
Really nice addition! I had some comments and nits etc, but the two big pieces to add would be docs updates and testing (where possible). I see, for example, that hf offers a way to emulate custom endpoints in the https://github.com/huggingface/hf-endpoints-emulator repo / package.

For docs, we'll need a new doc in the deployers section explaining exactly how to set this up on HF and how to use within ZenML.

Really excited about this one! Thanks for putting in the work!

safoinme reviewed (changes requested):
Thank you for all the work and the effort you've put in this PR. It's definitely a great addition to our deployers. I've got a couple of small suggestions.
Can't wait to see this released!

dudeperf3ct commented on src/zenml/integrations/huggingface/model_deployers/huggingface_model_deployer.py:
```diff
@@ -0,0 +1,440 @@
+            " If there's an active pipeline and/or model this artifact will be associated with it."
+        )
+
+        service_metadata = service.dict()
```
We use it to override UUID value and adding metadata while logging

```python
service_metadata = service.dict()
# UUID object is not json serializable
service_metadata["uuid"] = str(service_metadata["uuid"])

log_artifact_metadata(
    artifact_name=HUGGINGFACE_SERVICE_ARTIFACT,
    artifact_version=artifact_version,
    metadata={HUGGINGFACE_SERVICE_ARTIFACT: service_metadata},
)
```

dudeperf3ct replied:
You are right, seems redundant. I will remove it. Thanks for catching that.

dudeperf3ct commented on src/zenml/integrations/huggingface/model_deployers/huggingface_model_deployer.py:
```diff
@@ -0,0 +1,440 @@
+
+                # if endpoint is provisioned externally
+                # we do not have saved artifact for it.
+                except KeyError:
```
Added following logging error message here.

```python
logger.error(f"No key found for endpoint {endpoint.name} provisioned externally")
```

safoinme commented on src/zenml/integrations/huggingface/services/huggingface_deployment.py:
```diff
@@ -0,0 +1,227 @@
+        if self.hf_endpoint.url is not None:
+            logger.info("Huggingface inference endpoint successfully deployed.")
+        else:
+            logger.info("Failed to start huggingface inference endpoint service...")
```
Yes, you are right it should be `raise Exception(f"An unexpected error occurred while provisioning the Huggingface inference endpoint: {e}")`

strickvl commented:
@dudeperf3ct can you check the CI errors and update as appropriate? there's some linting issues and also some docstrings to be added as far as I can see.

dudeperf3ct commented:
@strickvl I have addressed the CI errors.

Re: Docs and Testing. I have added docs for the new deployer. I am not sure about using [https://github.com/huggingface/hf-endpoints-emulator](https://github.com/huggingface/hf-endpoints-emulator) repo / package for testing. 

Should I implement an example under [tests/integration/examples](https://github.com/zenml-io/zenml/tree/main/tests/integration/examples) for this new deployer? As part of running this step, an endpoint should be provisioned. Should I mock creating endpoint and use the emulated endpoint instead?

strickvl commented:
@dudeperf3ct yeah that was sort of my idea. Either way works IMO, but some kind of an end-to-end example for this would be great that we can test in the integration tests runners.

dudeperf3ct commented:
@strickvl  Re: CI error and dependencies. `huggingface_hub` dependency is added as part of `hugggingface` integration.

According to [docs](https://huggingface.co/docs/huggingface_hub/en/guides/inference_endpoints), the minimal version of `huggingface_hub` package supporting Inference Endpoints API is `v0.19.0`. 

Not sure what can be done to avoid the dependency conflict on `fsspec` package raised while setting up the environment.

![image](https://github.com/zenml-io/zenml/assets/10743098/170c4212-eb0c-461b-81c9-67a0f38a907e)

strickvl commented:
@dudeperf3ct that seems to be coming from the Azure dependency and `adlfs` which we have pinned down quite hard to one specific version (`2021.10.0`) via the Azure integration. It seems due a bump anyway. I'll work on freeing you up that way. Will make a PR but for now maybe just try https://github.com/zenml-io/zenml/blob/d216c662fb65e5abfaad6729b0f2d826f0611408/src/zenml/integrations/azure/__init__.py#L42-L43 as >= instead of ==?

strickvl commented:
Looks like S3FS will need to be bumped as well...

strickvl commented:
@dudeperf3ct #2402 looks on the surface of things to be ok, so maybe just add the change for S3FS into this PR and let's see if we can get it to work here too? (Hopefully will get that PR merged soon too, but in the interests of time...)

strickvl commented on docs/book/stacks-and-components/component-guide/model-deployers/huggingface.md:
Can you also add this new one to the `toc.md` in the appropriate place?

strickvl commented on docs/book/stacks-and-components/component-guide/model-deployers/huggingface.md:
```diff
@@ -0,0 +1,153 @@
+
+* if you want to deploy [Transformers, Sentence-Transformers, or Diffusion models](https://huggingface.co/docs/inference-endpoints/supported_tasks) on dedicated and secure infrastructure.
+* if you prefer a fully-managed production solution for inference without the need to handle containers and GPUs.
+* if your goal is to turn your models into production-ready APIs with minimal infrastructure or MLOps involvement * Cost-effectiveness is crucial, and you want to pay only for the raw compute resources you use.
```
```suggestion
* if your goal is to turn your models into production-ready APIs with minimal infrastructure or MLOps involvement 
* Cost-effectiveness is crucial, and you want to pay only for the raw compute resources you use.
```

strickvl commented on docs/book/stacks-and-components/component-guide/model-deployers/huggingface.md:
```diff
@@ -0,0 +1,153 @@
+
+## How to deploy it?
+
+The Huggingface Model Deployer flavor is provided by the Huggingface ZenML integration, so you need to install it on your local machine to be able to deploy your models. You can do this by running the following command:
```
```suggestion
The Hugging Face Model Deployer flavor is provided by the Hugging Face ZenML integration, so you need to install it on your local machine to be able to deploy your models. You can do this by running the following command:
```

strickvl commented on src/zenml/integrations/huggingface/services/__init__.py:
```diff
@@ -1,3 +1,16 @@
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
 """Initialization of the Huggingface Service."""
```
```suggestion
"""Initialization of the Hugging Face Service."""
```

strickvl commented on src/zenml/integrations/huggingface/services/huggingface_deployment.py:
```diff
@@ -148,23 +182,27 @@ def check_status(self) -> Tuple[ServiceState, str]:
                 "HuggingFace Inference Endpoint deployment is available",
```
```suggestion
                "Hugging Face Inference Endpoint deployment is available",
```

dudeperf3ct commented on docs/book/stacks-and-components/component-guide/model-deployers/huggingface.md:
```diff
@@ -0,0 +1,153 @@
+* `framework`: The machine learning framework used for the model (e.g. `"custom"`, `"pytorch"` )
+* `accelerator`: The hardware accelerator to be used for inference. (e.g. `"cpu"`, `"gpu"`)
+* `instance_size`: The size of the instance to be used for hosting the model (e.g. `"large"`, `"xxlarge"`)
+* `instance_type`: Inference Endpoints offers a selection of curated CPU and GPU instances. (e.g. `"c6i"`, `"g5.12xlarge"`)
```
There are no docs. I have included link to [code](https://github.com/huggingface/huggingface_hub/blob/5e3b603ccc7cd6523d998e75f82848215abf9415/src/huggingface_hub/hf_api.py#L6957) instead.

dudeperf3ct commented on docs/book/stacks-and-components/component-guide/model-deployers/huggingface.md:
````diff
@@ -0,0 +1,153 @@
+    predictions = predictor(model_deployment_service, inference_data)
+```
+
+For more information and a full list of configurable attributes of the Huggingface Model Deployer, check out
+the [SDK Docs]().
````
Not sure if the link is correct `https://sdkdocs.zenml.io/latest/integration_code_docs/integrations-huggingface/#zenml.integrations.huggingface.model_deployers`

dudeperf3ct commented on docs/book/stacks-and-components/component-guide/model-deployers/huggingface.md:
```diff
@@ -0,0 +1,153 @@
+* `endpoint_type`: (Optional) The type of the Inference Endpoint, which can be `"protected"`, `"public"` (default) or `"private"`.
+
+For more information and a full list of configurable attributes of the Huggingface Model Deployer, check out
+the [API Docs]().
```
Not sure if the link is correct https://sdkdocs.zenml.io/latest/integration_code_docs/integrations-huggingface/#zenml.integrations.huggingface.model_deployers

strickvl replied:
looks ok to me

strickvl commented on src/zenml/integrations/huggingface/services/huggingface_deployment.py:
```diff
@@ -0,0 +1,231 @@
+            return (
+                ServiceState.PENDING_STARTUP,
+                "HuggingFace Inference Endpoint deployment is being created: ",
+            )
```
for this I think we can probably implement the other statuses, no? Those all seem reasonable statuses to want to represent somehow. WDYT?

strickvl commented on src/zenml/integrations/huggingface/services/huggingface_deployment.py:
```diff
@@ -0,0 +1,231 @@
+            raise NotImplementedError(
+                "Tasks other than text-generation is not implemented."
+            )
+        return result
```
for this I would say I'd be fine with starting off with what we have and then we can add more tasks later on.

strickvl commented on pyproject.toml:
```diff
@@ -112,7 +112,7 @@ azure-storage-blob = { version = ">=12.0.0", optional = true }
 
 # Optional dependencies for the S3 artifact store
-s3fs = { version = "2022.11.0", optional = true }
+s3fs = { version = ">2022.3.0", optional = true }
```
```suggestion
s3fs = { version = ">2022.11.0", optional = true }
```

safoinme reviewed (approved):
Let's :shipit:
